### DISCUSSION TITLE
horovodrun vs. Python in multi-node manner

### DISCUSSION THREAD

**Discussion by assafna** [horovod/horovod#3582] [Q&A]
Hi all,
What are the differences between submitting a job with "horovodrun" command vs. with regular Python command?

When I tried to run a simple PyTorch multi-node [job](https://github.com/horovod/horovod/blob/master/examples/pytorch/pytorch_synthetic_benchmark.py) (2 nodes, with 2 GPUs on each) through Slurm (with a single process per GPU) using only Python it surprisingly worked and recognized 4 GPUs, so why should I use horovodrun anyway?

Also, how did it know to expect 2 nodes and 2 GPUs on each without me telling him? Is it thanks to Slurm?

Thanks,
Assaf.

**Comment by EnricoMi** [comment]
With `Python command`, you mean the Python API?

To answer you question correctly it would help you share the command line / Python API call.

**Reply by assafna** [reply]
Hi, thank you for your reply, the Python command is the one I referred to when I mentioned the simple PyTorch multi-node [job](https://github.com/horovod/horovod/blob/master/examples/pytorch/pytorch_synthetic_benchmark.py), it is launched without "horovodrun" and still works for multi-node.

`python pytorch_synthetic_benchmark.py`

**Reply by assafna** [reply]
@EnricoMi, do you need any other information?

**Reply by maxhgerlach** [reply]
@assafna, could you double check your SLURM jobs please? I have a hard time understanding how a script launched via `python pytorch_synthetic_benchmark.py` would be able to run distributed.

**Reply by assafna** [reply]
@maxhgerlach 

Some background:
I'm using a cluster with a few [NVIDIA DGX A100](https://www.nvidia.com/en-us/data-center/dgx-a100/) nodes, orchestrated with Slurm. The container is launched using [NVIDIA Pyxis](https://github.com/NVIDIA/pyxis) which uses a [squash file](https://en.wikipedia.org/wiki/SquashFS). To create a squash file, I used [NVIDIA Enroot](https://github.com/NVIDIA/enroot) to pull a [PyTorch container](https://catalog.ngc.nvidia.com/orgs/nvidia/containers/pytorch) from [NVIDIA NGC](https://catalog.ngc.nvidia.com/), install Horovod on it and export it to a squash file.

I'm attaching 4 files:
1. `example.py` - a very simple neural network example with a single forward and backward passes, with Horovod's code added to it.
2. `example.sub` - a Slurm's `sbatch` script file, allocating 2 nodes with a single GPU on each node, to enable a minimum configuration for multi-node.
3. `slurm-16161.out` - Slurm's output for this job.
4. `slurm-16162.out` - Slurm's output for Horovod's syntetic benchmark job.

`example.sub` contains a `srun` command which, using NVIDIA Pyxis, loads the container image and mounts the folder to access `example.py`, then it runs the file using python. I've started the `sbatch` script file with `sbatch example.sub` from a master node.

The same works with Horovod's GitHub `pytorch_synthetic_benchmark.py` file.

[horovod.zip](https://github.com/horovod/horovod/files/9175513/horovod.zip)

**Reply by maxhgerlach** [reply]
OK, so that's quite a bit more of a tech stack than the original post alluded to. I don't know what Pyxis does. Presumably, srun functions like mpirun in your case (for MPI, horovodrun is a wrapper around mpirun). It's been a while since I last worked with SLURM, so that might also be standard... 

You could set an environment variable like `HOROVOD_LOG_LEVEL=trace` to get more verbose logging and help you understand what's going on.

**Reply by maxhgerlach** [reply]
Yep, I wasn't aware about this: https://stackoverflow.com/questions/15578009/difference-between-running-a-program-with-and-without-mpirun

**Reply by assafna** [reply]
Interesting. So basically what you're saying is that Slurm's srun functions like mpirun (and like a "high-quality implementation"), and so it "gives" Horovod whatever it needs to recognize the "world" and the number of nodes and GPUs?

In conclusion, "horovodrun" is redundant when using Slurm?

Output with tracing enabled:
[slurm-16164.txt](https://github.com/horovod/horovod/files/9175548/slurm-16164.txt)

**Reply by maxhgerlach** [reply]
Thanks for the logs. They confirm that you are running with MPI (and that there's not some other magic enabling Gloo).

horovodrun can make it more convenient to set some configuration flags and there might be extra error handling, but if you don't need any of the extra features (like elastic mode, which wouldn't work with MPI), you don't absolutely need it.

**Reply by assafna** [reply]
Can you explain what happens first, is it that Slurm starts MPI and then Horovod wraps around it, or is it that Horovod starts and uses a backend of MPI? Or is it simpler than that, and just MPI being provided to Horovod which uses it as a backend?

Either way, the MPI backend is the one which allows Horovod to know the number of nodes and GPUs and enable multi-node runs automatically without "horovodrun"? I ask because when not using Horovod (E.g., PyTorch DDP) everything should be explicitly provided, such as the number of nodes, GPUs, ranks and so on (and also advised to be used with "torchrun"), even though it still uses a MPI backend. This is still "magic" to me when it come to Horovod.

BTW, up to this moment I wasn't able to run multi-node tasks with PT DDP and Slurm SBATCH, only with Horovod. But that another topic I'm trying to understand.
